RecoLocalTracker/SiStripClusterizer/plugins/ClustersFromZSRawProducer.cc

```diff
@@ -0,0 +1,300 @@
+
+      // get raw data
+      edm::Handle<FEDRawDataCollection> rawData;
+      ev.getByToken(productToken_, rawData);
+
+      auto output = std::make_unique<edmNew::DetSetVector<SiStripCluster>>();
+      assert(!output->onDemand());
+
+      output->reserve(15000, 32 * 10000);
+
+      run(*rawData, *output);
+      output->shrink_to_fit();
+      COUT << output->dataSize() << " clusters from " << output->size() << " modules" << std::endl;
+      ev.put(std::move(output));
+    }
+
+  private:
+    void initialize(const edm::EventSetup& es);
+
+    void run(const FEDRawDataCollection& rawColl, edmNew::DetSetVector<SiStripCluster>& output);
+
+  private:
+    edm::EDGetTokenT<FEDRawDataCollection> productToken_;
+
+    ClustersFromZSRawAlgorithm clusterizer_;
+
+    // March 2012: add flag for disabling APVe check in configuration
+    bool doAPVEmulatorCheck_;
+  };
+}  // namespace
+
+#include "FWCore/Framework/interface/MakerMacros.h"
+DEFINE_FWK_MODULE(SiStripClusterizerFromZSRaw);
+
+namespace {
+
+  void SiStripClusterizerFromZSRaw::initialize(const edm::EventSetup& es) { clusterizer_.initialize(es); }
+
+  void SiStripClusterizerFromZSRaw::run(const FEDRawDataCollection& rawColl,
+                                        edmNew::DetSetVector<SiStripCluster>& output) {
+    ClusterFiller filler(rawColl, clusterizer_, doAPVEmulatorCheck_);
+
+    // loop over good det in cabling
+    for (auto const& idet : clusterizer_.allDets()) {
+      StripClusterizerAlgorithm::output_t::FastFiller record(output, idet.detId);
+
+      filler.fill(idet, record);
+
+      if (record.empty())
+        record.abort();
+
+    }  // end loop over dets
+  }
+
+  void ClusterFiller::fill(StripClusterizerAlgorithm::Det const& det,
+                           StripClusterizerAlgorithm::output_t::FastFiller& record) {
+    incReady();
+
+    auto idet = record.id();
+
+    COUT << "filling " << idet << std::endl;
+
+    if (!det.valid())
+      return;
+
+    incSet();
+
+    // Loop over apv-pairs of det
+    for (auto const conn : clusterizer.currentConnection(det)) {
+      if
+        UNLIKELY(!conn) continue;
+
+      const uint16_t fedId = conn->fedId();
+
+      // If fed id is null or connection is invalid continue
+      if
+        UNLIKELY(!fedId || !conn->isConnected()) { continue; }
+
+      // If Fed hasnt already been initialised, extract data and initialise
+      if (!buffers[fedId])
+        buffers[fedId] = fillBuffer(fedId, rawColl);
+      if (!buffers[fedId])
+        continue;
+      auto buffer = buffers[fedId].get();
+
+      buffer->setLegacyMode(false);
+
+      // check channel
+      const uint8_t fedCh = conn->fedCh();
+
+      if
+        UNLIKELY(!buffer->channelGood(fedCh, doAPVEmulatorCheck)) {
+          if (edm::isDebugEnabled()) {
+            std::ostringstream ss;
+            ss << "Problem unpacking channel " << fedCh << " on FED " << fedId;
+            edm::LogWarning("ClustersFromZSRawProducer") << ss.str();
```

Comment:
```sistrip::mlRawToCluster_``` category?
also, the intermediate ostringstream instance looks unnecessary

Reply:
old code.